### DISCUSSION TITLE
how to get final update from notifier via async iterable?

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#5924] [Q&A] [answered]
At office hours, somebody reports that they export a notifier from a contract; when they consume it using `makeAsyncIterableFromNotifier`, they don't get the value from `updater.finish()`.

I tried to reduce it to a smallish test case, but I'm struggling. The following doesn't seem to get any updates at all out of the iterator. It prints none of the newspaper headlines; just

```
stpe1
step3
step2
```

The code I tried is:

```js
// @ts-check

// import { test } from '@agoric/zoe/tools/prepare-test-env-ava.js';
import '@endo/init';

import {
  makeNotifierKit,
  makeAsyncIterableFromNotifier,
} from '@agoric/notifier';

// test('getting final result', async t => {
// });

const trial = async () => {
  const newspaper = makeNotifierKit();

  const publishNewsPaper = async () => {
    newspaper.updater.updateState('Aug 1');
    newspaper.updater.updateState('Aug 2');
    newspaper.updater.updateState('Aug 3');
    newspaper.updater.finish('Nov 2');
    console.log('step3');
  };

  const readNewspaper = async () => {
    const iter = makeAsyncIterableFromNotifier(newspaper.notifier);

    console.log('step2');

    for await (const headlines of iter) {
      console.log(headlines);
    }
  };

  console.log('stpe1');
  await Promise.all([publishNewsPaper(), readNewspaper()]);
};

trial().catch(err => console.error(err));
```

cc @kriskowal

**Comment by dckc** [comment]
The dev pointed out that our dapps seem to rely on getting the final value:

https://github.com/Agoric/dapp-treasury/blob/4aa1fc717a593be5f39cc74a6206710209c8f377/ui/src/contexts/Application.jsx#L78-L85

https://github.com/Agoric/agoric-sdk/blob/6f72d00dddf3467fca0b70e1455e19f5d87adea8/packages/inter-protocol/src/vaultFactory/vault.js#L364-L371

or perhaps not?

**Comment by anilhelvaci** [comment]
Hello everybody 👋 I'm the dev from who asked this question.

I think I made Dan's code work by simply waiting for one second after every state change. This worked because as far as I know notifier only pushes latest state and if a newer state arrives before it pushed state at hand, it pushes newest one and ignores the rest. The code looks like this;

```js
// @ts-check

import '@endo/init';
import {
  makeNotifierKit,
  makeAsyncIterableFromNotifier,
} from '@agoric/notifier';

const trial = async () => {
  const newspaper = makeNotifierKit();

  const publishNewsPaper = async () => {
    newspaper.updater.updateState('Aug 1');
    await new Promise(resolve => setTimeout(resolve, 1000));
    newspaper.updater.updateState('Aug 2');
    await new Promise(resolve => setTimeout(resolve, 1000));
    newspaper.updater.updateState('Aug 3');
    await new Promise(resolve => setTimeout(resolve, 1000));
    newspaper.updater.finish('Nov 2');
    console.log('step3');
  };

  const readNewspaper = async () => {
    const iter = makeAsyncIterableFromNotifier(newspaper.notifier);

    console.log('step2');

    for await (const headlines of iter) {
      console.log(headlines);
    }
  };

  console.log('step1');
  await Promise.all([publishNewsPaper(), readNewspaper()]);
};

trial().catch(err => console.error(err));
```

and the output

```
step1
step2
Aug 1
Aug 2
Aug 3
step3
```

Basically yes, we did not get the `finalState` inside for/await/of loop in this test as well. So;
* Is this how it should work?
* If it is, getting the final state via `getUpdateSince` under the for/await/of loop is a considerable approach or not? 
     * If it's not, how should I get the final state?

**Comment by kriskowal** [comment, marked as answer]
The aync iterator is working as expected. JavaScript’s [iterator protocols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) support a final value but `for` loops do not surface it. There is a possibility that `for` expressions will eventually be possible in the language, such that the following would surface that final value:

```js
const final = await for await (const medial of iterable) {}
```

But, today, to see the final value, you have to implement the iterator protocol by hand:

```js
async function forAwaitEach(iterable, iterback)  {
  const iterator = iterator[Symbol.asyncIterator]();
  try {
    for (;;) {
      const {done, value} = await iterator.next();
      if (done) {
        return value; // <- this is the final value that for loops don’t surface
      }
      await iterback(value);
    }
  } finally {
    if (iterator.return) {
      iterator.return();
    }
  }
}

const final = await forAwaitEach(iterator, medial => {});
```
